mesonbuild/compilers/compilers.py

```diff
@@ -544,70 +585,125 @@ def get_linker_args_from_envvars(self,
+                 dependencies: T.Optional[T.List['Dependency']] = None,
+                 mode: str = 'compile',
+                 disable_cache: bool = False) -> T.Tuple[bool, bool]:
         raise EnvironmentException('Language %s does not support compile checks.' % self.get_display_language())
 
-    def links(self, *args, **kwargs) -> T.Tuple[bool, bool]:
+    def links(self, code: str, env: 'Environment', *,
+              extra_args: T.Union[None, T.List[str], CompilerArgs] = None,
+              dependencies: T.Optional[T.List['Dependency']] = None,
+              mode: str = 'compile',
+              disable_cache: bool = False) -> T.Tuple[bool, bool]:
         raise EnvironmentException('Language %s does not support link checks.' % self.get_display_language())
 
-    def run(self, *args, **kwargs) -> RunResult:
+    def run(self, code: str, env: 'Environment', *,
+            extra_args: T.Optional[T.List[str]] = None,
+            dependencies: T.Optional[T.List['Dependency']] = None) -> RunResult:
         raise EnvironmentException('Language %s does not support run checks.' % self.get_display_language())
 
-    def sizeof(self, *args, **kwargs) -> int:
+    def sizeof(self, typename: str, prefix: str, env: 'Environment', *,
+               extra_args: T.Optional[T.List[str]] = None,
+               dependencies: T.Optional[T.List['Dependency']] = None) -> int:
         raise EnvironmentException('Language %s does not support sizeof checks.' % self.get_display_language())
 
-    def alignment(self, *args, **kwargs) -> int:
+    def alignment(self, typename: str, prefix: str, env: 'Environment', *,
+                 extra_args: T.Optional[T.List[str]] = None,
+                 dependencies: T.Optional[T.List['Dependency']] = None) -> int:
         raise EnvironmentException('Language %s does not support alignment checks.' % self.get_display_language())
 
-    def has_function(self, *args, **kwargs) -> T.Tuple[bool, bool]:
+    def has_function(self, typename: str, prefix: str, env: 'Environment', *,
+                     extra_args: T.Optional[T.Iterable[str]] = None,
+                     dependencies: T.Optional[T.Iterable['Dependency']] = None) -> T.Tuple[bool, bool]:
+        """See if a function exists.
+
+        Returns a two item tuple of bools. The first bool is whether the
+        check succeeded, the second is whether the result was cached (True)
+        or run fresh (False).
+        """
         raise EnvironmentException('Language %s does not support function checks.' % self.get_display_language())
 
     @classmethod
     def unix_args_to_native(cls, args: T.List[str]) -> T.List[str]:
         "Always returns a copy that can be independently mutated"
-        return args[:]
+        return args.copy()
 
     @classmethod
     def native_args_to_unix(cls, args: T.List[str]) -> T.List[str]:
         "Always returns a copy that can be independently mutated"
-        return args[:]
+        return args.copy()
```

Comment:
Is there a specific reason for switching to `.copy()`? According to [StackOverflow](https://stackoverflow.com/questions/2612802/list-changes-unexpectedly-after-assignment-how-do-i-clone-or-copy-it-to-prevent) the old slicing should be faster.

Reply:
I think arg conversion is also a hotspot, so the speed difference may be significant.

Reply:
Python is my primary programming language, and I had to look up with the heck `[:]` does. It's extremely non-obvious, but anyone can see `foo.copy()` and reasonably understand what that does.

I'm suspicious of those benchmarking numbers, especially because python 3 has a bunch of optimizations to make using methods faster that python2 lacked. And that answer doesn't use `list.copy()`, but `copy.copy`, since the former was added in python3, and not present in python2.

```python
from timeit import timeit

a = [str(a) for a in range(50)]

print('slice:', timeit(lambda: a[:]))
print('copy:', timeit(lambda: a.copy()))
print('new list:', timeit(lambda: list(a)))
```
and here's my results:

python3.6:
```
slice: 0.3625514349998866
copy: 0.36979652300010457
new list: 0.4572163679999903
```

python3.7
```
slice: 0.346616264999966
copy: 0.3320672290001312
new list: 0.41092001400011213
```

python3.8
```
slice: 0.30240244600008737
copy: 0.2990029269999468
new list: 0.319297583999969
```

python3.9
```
slice: 0.3030145709999488
copy: 0.26365815300005124
new list: 0.3073842159999458
```

Note that timeit calls the callable 1,000,000 times. The difference in performance between copy() and slicing is minimal (and varies, sometimes one is faster than the other.)

Reply:
Thanks for doing the testing, lgtm :)